File: sf/models/SfTypeMap.py
from __future__ import annotations
import datetime, json, re
from decimal import Decimal
from typing import Any, Literal, TypeVar
from server.plugins.PluginModels import Catalog, Entity, Column, ArrowReader, Records, arrow_type_literal, pa_type_to_literal
import pyarrow as pa

SF_TYPE_TO_ARROW = {
    "string":          pa.utf8(),
    "textarea":        pa.large_utf8(),
    "phone":           pa.utf8(),
    "email":           pa.utf8(),
    "url":             pa.utf8(),
    "picklist":        pa.utf8(),       # or pa.dictionary if cardinality is low
    "multipicklist":   pa.large_utf8(), # semicolon-separated, SF's crime against data
    "combobox":        pa.utf8(),
    "id":              pa.utf8(),       # SF IDs are 18-char strings
    "reference":       pa.utf8(),       # FK - SF ID stored as string, metadata captured separately
    "boolean":         pa.bool_(),
    "int":             pa.int32(),
    "long":            pa.int64(),
    "double":          pa.float64(),
    "currency":        pa.decimal128(18, 2),  # don't use float for money
    "percent":         pa.float64(),
    "date":            pa.date32(),
    "datetime":        pa.timestamp("ms", tz="UTC"),  # SF datetimes are always UTC
    "time":            pa.time64("us"),
    "base64":          pa.large_binary(),
    "encryptedstring": pa.utf8(),       # SF-side encrypted, comes to you as string
    "anyType":         pa.utf8(),       # fallback, SF uses this for polymorphic value fields
    "address":         None,            # compound - exclude at describe time
    "location":        None,            # compound geolocation - exclude
}

# ---------------------------------------------------------------
# SF type → arrow_type_literal  (Column.arrow_type_id)
# ---------------------------------------------------------------
SF_TO_ARROW_LITERAL: dict[str, arrow_type_literal] = {
    "string":          "utf8",
    "textarea":        "large_string",
    "phone":           "utf8",
    "email":           "utf8",
    "url":             "utf8",
    "picklist":        "utf8",
    "multipicklist":   "large_string",
    "combobox":        "utf8",
    "id":              "utf8",
    "reference":       "utf8",
    "boolean":         "bool",
    "int":             "int32",
    "long":            "int64",
    "double":          "float64",
    "currency":        "decimal128",
    "percent":         "float64",
    "date":            "date32",
    "datetime":        "timestamp_ms",
    "time":            "time64_us",
    "base64":          "large_binary",
    "encryptedstring": "utf8",
    "anyType":         "utf8",
}

# FieldDefinition.DataType (bulk describe) → PythonTypes
# DataType comes with optional params e.g. "Text(80)", "Number(18, 0)" - strip before lookup.
def _normalize_fielddef_type(raw: str) -> str:
    """Strip parameters from FieldDefinition DataType strings.
    e.g. 'Text(80)' -> 'text', 'Number(18, 0)' -> 'number'
    """
    return re.sub(r'\(.*\)', '', raw).strip().lower()

# Legacy mapping.
PythonTypes = Literal[
    "string",
    "integer",
    "float",
    "boolean",
    "datetime", # datetime.datetime # timezone format
    "date",     # datetime.date
    "time",     # datetime.time
    "byte",
    "bytearray",
    "json",     # dict or list
]

SF_FIELDDEF_TYPE_MAP: dict[str, PythonTypes] = {
    'id':                    'string',
    'text':                  'string',
    'textarea':              'string',
    'longtextarea':          'string',
    'html':                  'string',
    'richtextarea':          'string',
    'phone':                 'string',
    'url':                   'string',
    'email':                 'string',
    'picklist':              'string',
    'multi-select picklist': 'string',
    'combobox':              'string',
    'reference':             'string',
    'autonumber':            'string',
    'encryptedtext':         'string',
    'hierarchy':             'string',
    'anytype':               'string',
    'number':                'float',
    'currency':              'float',
    'percent':               'float',
    'double':                'float',
    'integer':               'integer',
    'long':                  'integer',
    'checkbox':              'boolean',
    'boolean':               'boolean',
    'date':                  'date',
    'date/time':             'datetime',
    'datetime':              'datetime',
    'time':                  'time',
    'base64':                'byte',
    'file':                  'byte',
    'json':                  'json',
    # Compound - filtered upstream before this map is consulted
    'address':               'json',
    'location':              'json',
}

# Individual SObject describe response `type` field → PythonTypes
SF_TYPE_MAP: dict[str, PythonTypes] = {
    'id':              'string',
    'string':          'string',
    'textarea':        'string',
    'email':           'string',
    'phone':           'string',
    'url':             'string',
    'encryptedstring': 'string',
    'picklist':        'string',
    'multipicklist':   'string',
    'combobox':        'string',
    'reference':       'string',
    'anyType':         'string',
    'int':             'integer',
    'integer':         'integer',
    'double':          'float',
    'currency':        'float',
    'percent':         'float',
    'boolean':         'boolean',
    'date':            'date',
    'datetime':        'datetime',
    'time':            'time',
    'base64':          'byte',
    'address':         'json',
    'location':        'json',
}


def _to_bool(v: Any) -> bool:
    if isinstance(v, bool):
        return v
    return str(v).lower() == 'true'


def _to_datetime(v: str) -> datetime.datetime:
    if v.endswith('Z'):
        v = v[:-1] + '+00:00'
    elif len(v) > 5 and v[-5] in '+-' and ':' not in v[-5:]:
        v = v[:-2] + ':' + v[-2:]
    return datetime.datetime.fromisoformat(v)


def _to_time(v: str) -> datetime.time:
    # Salesforce time fields use ISO-8601 with a trailing 'Z' (e.g. '00:00:00.000Z')
    # which datetime.time.fromisoformat() rejects. Strip it before parsing.
    if isinstance(v, str) and v.endswith('Z'):
        v = v[:-1]
    return datetime.time.fromisoformat(v)


ARROW_TO_SF_TYPE: dict[arrow_type_literal, str] = {
    "null":         "anyType",
    "bool":         "boolean",
    "int8":         "int",
    "int16":        "int",
    "int32":        "int",
    "int64":        "long",
    "uint8":        "int",
    "uint16":       "int",
    "uint32":       "long",
    "uint64":       "long",
    "float16":      "double",
    "float32":      "double",
    "float64":      "double",
    "decimal128":   "currency",
    "decimal256":   "currency",
    "string":       "string",
    "utf8":         "string",
    "large_string": "textarea",
    "string_view":  "string",
    "binary":       "base64",
    "large_binary": "base64",
    "date32":       "date",
    "date64":       "date",
    "timestamp_s":  "datetime",
    "timestamp_ms": "datetime",
    "timestamp_us": "datetime",
    "timestamp_ns": "datetime",
    "time32_s":     "time",
    "time32_ms":    "time",
    "time64_us":    "time",
    "time64_ns":    "time",
    "duration_s":   "string",
    "duration_ms":  "string",
    "duration_us":  "string",
    "duration_ns":  "string",
    "list":         "multipicklist",
    "large_list":   "multipicklist",
    "list_view":    "multipicklist",
    "large_list_view": "multipicklist",
    "struct":       "anyType",
    "map":          "anyType",
    "dictionary":   "picklist",
    "json":         "anyType",
    "uuid":         "string",
}


def arrow_to_sf_type(t: pa.DataType) -> str:
    """Map any pa.DataType to its closest Salesforce field type string."""
    return ARROW_TO_SF_TYPE.get(pa_type_to_literal(t), "anyType")


def _to_decimal(v: Any) -> Decimal:
    # str() avoids floating-point representation noise (e.g. 1234.56 not 1234.5599999...)
    return Decimal(str(v))


_SF_CONVERTERS: dict[str, Any] = {
    'int':      int,
    'integer':  int,
    'double':   float,
    'currency': _to_decimal,  # decimal128 requires Decimal, not float
    'percent':  float,
    'boolean':  _to_bool,
    'date':     datetime.date.fromisoformat,
    'datetime': _to_datetime,
    'time':     _to_time,
}


def sf_to_python(sf_type: str, value: Any) -> Any:
    """Convert a Salesforce field value to its native Python type."""
    if value is None or value == '': return None
    converter = _SF_CONVERTERS.get(sf_type)
    if converter:
        try: return converter(value)
        except (ValueError, TypeError): return value
    return value


def python_to_sf(value: Any) -> str:
    """Convert a Python value to its Salesforce API string representation."""
    if value is None: return ''
    if isinstance(value, bool): return 'true' if value else 'false'
    if isinstance(value, datetime.datetime): return value.isoformat()
    if isinstance(value, datetime.date): return value.isoformat()
    if isinstance(value, datetime.time): return value.isoformat()
    if isinstance(value, (dict, list)): return json.dumps(value)
    return str(value)

def cast_record(record: dict[str, Any], field_types: dict[str, str]) -> dict[str, Any]:
    """Apply sf_to_python to each field in a record using a {field_name: sf_type} map."""
    return {
        k: sf_to_python(field_types[k], v) if k in field_types else v
        for k, v in record.items()
    }

_CLEAR = object()  # sentinel for explicit null writes to SF

def prepare_record(record: dict[str, Any]) -> dict[str, Any]:
    """Convert Python values to SF API representation.
    Use _CLEAR as a value to explicitly null a field in SF.
    Omit a key entirely to leave the field unchanged.
    """
    out = {}
    for k, v in record.items():
        if v is _CLEAR: out[k] = None  # explicit null - SF will clear the field
        elif v is not None: out[k] = python_to_sf(v)
    return out


def date_to_iso8601(date: datetime.date) -> str:
    """Returns an ISO8601 string from a date"""
    datetimestr = date.strftime('%Y-%m-%dT%H:%M:%S')
    timezonestr = date.strftime('%z')
    return (
        f'{datetimestr}{timezonestr[0:3]}:{timezonestr[3:5]}'
        .replace(':', '%3A')
        .replace('+', '%2B')
    )

File: typescripts/src/sf/SfTypeMap.ts
/**
 * SfTypeMap.ts
 *
 * Salesforce field-type mapping and value conversion, ported from
 * src/sf/SfTypeMap.py. JS has no Decimal/date/time types, so: currency is kept
 * as a numeric string (precision-preserving for the downstream Oracle NUMBER
 * bind), date/datetime become Date, and time stays a string (Oracle stores time
 * as VARCHAR2).
 */
import { PythonTypes, type Row } from '../models.js'

export const SF_TYPE_MAP: Record<string, PythonTypes> = {
  id: PythonTypes.string,
  string: PythonTypes.string,
  textarea: PythonTypes.string,
  email: PythonTypes.string,
  phone: PythonTypes.string,
  url: PythonTypes.string,
  encryptedstring: PythonTypes.string,
  picklist: PythonTypes.string,
  multipicklist: PythonTypes.string,
  combobox: PythonTypes.string,
  reference: PythonTypes.string,
  anytype: PythonTypes.string,
  int: PythonTypes.integer,
  integer: PythonTypes.integer,
  long: PythonTypes.integer,
  double: PythonTypes.float,
  currency: PythonTypes.float,
  percent: PythonTypes.float,
  boolean: PythonTypes.boolean,
  date: PythonTypes.date,
  datetime: PythonTypes.datetime,
  time: PythonTypes.time,
  base64: PythonTypes.byte,
  complexvalue: PythonTypes.json,
  address: PythonTypes.json,
  location: PythonTypes.json,
}

/**
 * Map a Salesforce describe field `type` to a PythonType. Unknown types fall
 * back to string (with a warning) so a single exotic field can't abort a full
 * schema migration.
 */
export function sfTypeToPython(sfType: string): PythonTypes {
  const mapped = SF_TYPE_MAP[sfType.toLowerCase()]
  if (mapped === undefined) {
    console.warn("Unmapped Salesforce field type '%s'; defaulting to string.", sfType)
    return PythonTypes.string
  }
  return mapped
}

function toBool(v: unknown): boolean {
  if (typeof v === 'boolean') return v
  return String(v).toLowerCase() === 'true'
}

function toDatetime(v: string): Date {
  let s = v
  if (s.endsWith('Z')) {
    s = s.slice(0, -1) + '+00:00'
  } else if (s.length > 5 && (s[s.length - 5] === '+' || s[s.length - 5] === '-') && !s.slice(-5).includes(':')) {
    s = s.slice(0, -2) + ':' + s.slice(-2)
  }
  const d = new Date(s)
  if (Number.isNaN(d.getTime())) throw new Error(`invalid datetime: ${v}`)
  return d
}

function toDateOnly(v: string): Date {
  const m = v.match(/^(\d{4})-(\d{2})-(\d{2})/)
  if (!m) throw new Error(`invalid date: ${v}`)
  return new Date(Date.UTC(Number(m[1]), Number(m[2]) - 1, Number(m[3])))
}

function toTime(v: string): string {
  return typeof v === 'string' && v.endsWith('Z') ? v.slice(0, -1) : v
}

const SF_CONVERTERS: Record<string, (v: unknown) => unknown> = {
  int: (v) => {
    const n = Number(v)
    if (!Number.isInteger(n)) throw new Error('not int')
    return n
  },
  integer: (v) => {
    const n = Number(v)
    if (!Number.isInteger(n)) throw new Error('not int')
    return n
  },
  long: (v) => {
    const n = Number(v)
    if (!Number.isInteger(n)) throw new Error('not int')
    return n
  },
  double: (v) => {
    const n = Number(v)
    if (Number.isNaN(n)) throw new Error('not a number')
    return n
  },
  percent: (v) => {
    const n = Number(v)
    if (Number.isNaN(n)) throw new Error('not a number')
    return n
  },
  // Keep currency as a numeric string to preserve precision (Decimal in Python).
  currency: (v) => String(v),
  boolean: toBool,
  date: (v) => toDateOnly(String(v)),
  datetime: (v) => toDatetime(String(v)),
  time: (v) => toTime(String(v)),
}

/** Convert a Salesforce field value to its native JS type. */
export function sfToPython(sfType: string, value: unknown): unknown {
  if (value === null || value === undefined || value === '') return null
  const converter = SF_CONVERTERS[sfType]
  if (converter) {
    try {
      return converter(value)
    } catch {
      return value
    }
  }
  return value
}

/** Apply sfToPython to each field using a {fieldName: sfType} map. */
export function castRecord(record: Row, fieldTypes: Record<string, string>): Row {
  const out: Row = {}
  for (const [k, v] of Object.entries(record)) {
    out[k] = k in fieldTypes ? sfToPython(fieldTypes[k], v) : v
  }
  return out
}

/** Sentinel for explicit null writes to Salesforce. */
export const CLEAR = Symbol('CLEAR')

/** Convert a JS value to its Salesforce API string representation. */
export function pythonToSf(value: unknown): string {
  if (value === null || value === undefined) return ''
  if (typeof value === 'boolean') return value ? 'true' : 'false'
  if (value instanceof Date) return value.toISOString()
  if (typeof value === 'object') return JSON.stringify(value)
  return String(value)
}

/**
 * Convert JS values to SF API representation. Use CLEAR to explicitly null a
 * field; omit a key entirely to leave the field unchanged.
 */
export function prepareRecord(record: Record<string, unknown>): Row {
  const out: Row = {}
  for (const [k, v] of Object.entries(record)) {
    if (v === CLEAR) {
      out[k] = null
    } else if (v !== null && v !== undefined) {
      out[k] = pythonToSf(v)
    }
  }
  return out
}
